File: 0285-Inorder Successor in BST/0285-Inorder Successor in BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
        TreeNode* succ = nullptr;
        while (root)
        {
            if (root->val <= p->val)
            {
                root = root->right;
            }
            else
            {
                succ = root;
                root = root->left;
            }
        }
        
        return succ;
    }
};

File: 0201-0300/0285-Inorder Successor in BST/0285-Inorder Successor in BST.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def inorderSuccessor(self, root: 'TreeNode', p: 'TreeNode') -> 'TreeNode':
        result = None
        while root:
            if root.val <= p.val:
                root = root.right
            else:
                result = root
                root = root.left
        return result
